presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java

```diff
@@ -58,62 +51,136 @@ public Block distinct(@TypeParameter("E") Type type, @SqlType("array(E)") Block
-            pageBuilder.reset();
-        }
+        if (array.mayHaveNull()) {
+            // Keep adding the element to the set as long as there are no dupes.
+            while (position < arrayLength && typedSet.add(array, position)) {
+                position++;
+            }
 
-        BlockBuilder distinctElementBlockBuilder = pageBuilder.getBlockBuilder(0);
-        for (int i = 0; i < array.getPositionCount(); i++) {
-            if (!typedSet.contains(array, i)) {
-                typedSet.add(array, i);
-                distinctCount++;
+            if (position == arrayLength) {
+                // All elements are distinct, so just return the original.
+                return array;
+            }
+
+            distinctElementBlockBuilder = type.createBlockBuilder(null, arrayLength);
+            for (int i = 0; i < position; i++) {
                 type.appendTo(array, i, distinctElementBlockBuilder);
             }
+            for (position++; position < arrayLength; position++) {
+                if (typedSet.add(array, position)) {
+                    type.appendTo(array, position, distinctElementBlockBuilder);
+                }
+            }
         }
+        else {
+            while (position < arrayLength && typedSet.addNonNull(array, position)) {
+                position++;
+            }
 
-        pageBuilder.declarePositions(distinctCount);
+            if (position == arrayLength) {
+                // All elements are distinct, so just return the original.
+                return array;
+            }
+
+            distinctElementBlockBuilder = type.createBlockBuilder(null, arrayLength);
+            for (int i = 0; i < position; i++) {
+                type.appendTo(array, i, distinctElementBlockBuilder);
+            }
+            for (position++; position < arrayLength; position++) {
+                if (typedSet.addNonNull(array, position)) {
+                    type.appendTo(array, position, distinctElementBlockBuilder);
+                }
+            }
+        }
 
-        return distinctElementBlockBuilder.getRegion(distinctElementBlockBuilder.getPositionCount() - distinctCount, distinctCount);
+        return distinctElementBlockBuilder.build();
     }
 
     @SqlType("array(bigint)")
-    public Block bigintDistinct(@SqlType("array(bigint)") Block array)
+    public static Block bigintDistinct(@SqlType("array(bigint)") Block array)
     {
-        if (array.getPositionCount() == 0) {
+        final int arrayLength = array.getPositionCount();
+        if (arrayLength < 2) {
             return array;
         }
 
-        boolean containsNull = false;
-        LongSet set = new LongOpenHashSet(array.getPositionCount());
-        int distinctCount = 0;
+        BlockBuilder distinctElementBlockBuilder;
+        LongSet set = new LongOpenHashSet(arrayLength);
+
+        if (array.mayHaveNull()) {
+            int position = 0;
+            boolean notSeenNull = true;
```

Comment:
I like the previous `containsNull` variable name better.

Reply:
Done

Reply:
Refactored as there was a test failure.